src/chess/bitboard.h

```diff
@@ -190,40 +190,64 @@ class Move {
  public:
   enum class Promotion : std::uint8_t { None, Queen, Rook, Bishop, Knight };
   Move() = default;
-  Move(BoardSquare from, BoardSquare to) : from_(from), to_(to) {}
-  Move(BoardSquare from, BoardSquare to, Promotion promotion)
-      : from_(from), to_(to), promotion_(promotion) {}
-  Move(const std::string& str, bool black = false);
-  Move(const char* str, bool black = false) : Move(std::string(str), black) {}
-
-  BoardSquare from() const { return from_; }
-  BoardSquare to() const { return to_; }
-  Promotion promotion() const { return promotion_; }
-  bool IsCastling() const { return castling_; }
-  void SetCastling() { castling_ = true; }
+  Move(BoardSquare from, BoardSquare to) {
+    SetTo(to);
+    SetFrom(from);
+  }
+  Move(BoardSquare from, BoardSquare to, Promotion promotion) {
+    SetTo(to);
+    SetFrom(from);
+    SetPromotion(promotion);
+  }
+  Move(const std::string &str, bool black = false);
+  Move(const char *str, bool black = false) : Move(std::string(str), black) {}
+
+  BoardSquare GetTo() const { return (BoardSquare)((data_ & TO_MASK_)); }
+  BoardSquare GetFrom() const {
+    return (BoardSquare)((data_ & FROM_MASK_) >> 6);
+  }
+  Promotion GetPromotion() const {
+    return (Promotion)((data_ & PROMO_MASK_) >> 12);
+  }
+  bool IsCastling() const { return (bool)((data_ & CASTLE_MASK_) >> 15); }
+  void SetCastling() { data_ |= CASTLE_MASK_; }
+
+  void SetTo(BoardSquare to) { data_ = (data_ & ~TO_MASK_) | to.as_int(); }
+  void SetFrom(BoardSquare from) {
+    data_ = (data_ & ~FROM_MASK_) | (from.as_int() << 6);
+  }
+  void SetPromotion(Promotion promotion) {
+    data_ = (data_ & ~PROMO_MASK_) | ((uint8_t)promotion << 12);
+  }
 
   // 0 .. 16384, knight promotion and no promotion is the same.
   uint16_t as_packed_int() const;
 
   // 0 .. 1857, to use in neural networks.
   uint16_t as_nn_index() const;
 
-  bool operator==(const Move& other) const {
-    return from_ == other.from_ && to_ == other.to_ &&
-           promotion_ == other.promotion_;
+  bool operator==(const Move &other) const {
+    return GetFrom() == other.GetFrom() && GetTo() == other.GetTo() &&
```

Comment:
Indeed it should be `data_ == other.data_`, castlings are expected to be different moves.

Reply:
Done